File: BT-Right-View.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int> result;
public:
    vector<int> rightSideView(TreeNode* root) {
        if(root==NULL){
            return result;
        }
        dfs(root,0);
        return result;
    }
    void dfs(TreeNode* root, int lvl){
        if(root==NULL){
            return;
        }
        if(result.size()==lvl){
            result.push_back(root->val);
        }
        dfs(root->right, lvl+1);
        dfs(root->left, lvl+1);
    }
};

File: bt_right_view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// dfs traversal
// Time Complexity : O(N)
// Space Complexity : O(h) h is height of recursive stack
// Did this code successfully run on Leetcode : Yes
class Solution {
    List<Integer> result = new ArrayList<>();
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null)
            return result;
        dfs(root, 0);        
        return result;
    }
    private void dfs(TreeNode root, int level){
        if(root == null)
            return;
        if(level == result.size()){
            result.add(root.val);
        }
        dfs(root.right, level+1);
        dfs(root.left, level+1);
        return;
    }
}

/*
// bfs traversal
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null)
            return result;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            for(int i=0; i<size; i++){
                TreeNode curr = q.poll();
                if(i == size-1)
                    result.add(curr.val);
                if(curr.left != null)
                    q.add(curr.left);
                if(curr.right != null)
                    q.add(curr.right);
            }
        }
        return result;
    }
}
*/